Add blog post about forensic container checkpointing (#37412)
kubernetes/website: content/en/blog/_posts/2022-10-31-forensic-container-checkpointing/index.md

uses these existing CRIU integrations.

[criu]: https://criu.org/

## Why is it important?

With the help of CRIU and the corresponding integrations it is possible to get
all information and state about a running container on disk for later forensic
analysis. Forensic analysis might be important to inspect a suspicious
container without stopping or influencing it. If the container is really under
attack, the attacker might detect attempts to inspect the container. Taking a
checkpoint and analysing the container in a sandboxed environment offers the
possibility to inspect the container without the original container and maybe
attacker being aware of the inspection.

In addition to the forensic container checkpointing use case, it is also
possible to migrate a container from one node to another node without loosing
the internal state. Especially for stateful containers with long initialization
times restoring from a checkpoint might save time after a reboot or enable much
faster startup times.

## How do I use container checkpointing?

The feature is behind a feature gate, so make sure to enable the
`ContainerCheckpointRestore` gate before you can use the new feature.

> The runtime must also support forensic container checkpointing:

Reviewer: ```suggestion
The runtime must also support container checkpointing:
```
The runtime is agnostic about the use to which you put the checkpoint feature.

* containerd: support is currently under discussion. See containerd
  pull request [#6965][containerd-checkpoint-restore-pr] for more details.

* CRI-O: v1.25 has support for forensic container checkpointing.

[containerd-checkpoint-restore-pr]: https://github.com/containerd/containerd/pull/6965

### Usage example with CRI-O

To use forensic container checkpointing in combination with CRI-O, CRI-O needs

> to be started with the command-line options `--enable-criu-support=true
> --drop-infra-ctr=false` and Kubernetes needs to be started with the

Reviewer: ```suggestion
to be started with the command-line options `--enable-criu-support=true` and
`--drop-infra-ctr=false`. For Kubernetes, you need to run your cluster with the
```

`ContainerCheckpointRestore` feature gate enabled. As the checkpointing
functionality is provided by CRIU it is also necessary to install CRIU.

> Usually runc or crun depend on CRIU and therefore it is installed automatically.

Reviewer: This is where I'd mention that the security implications are still under consideration (it's OK to repeat that mention later).

Author: added

Once containers and pods are running it is possible to create a checkpoint.
Checkpointing is currently only exposed on the **kubelet** level. To checkpoint
a container, you can run `curl` on the node where that container is running,
and trigger a checkpoint:

```shell
curl -X POST "https://localhost:10250/checkpoint/namespace/podId/container"
```

For a container named *counter* in a pod named *counters* in a namespace named
*default* the **kubelet** API endpoint is reachable at:

```shell
curl -X POST "https://localhost:10250/checkpoint/default/counters/counter"
```

For completeness the following `curl` command-line options are necessary to
have `curl` trust the kubelet's CA certificate:

```shell

> -k --cert /var/run/kubernetes/client-admin.crt --key /var/run/kubernetes/client-admin.key

Reviewer: `-k` does not _trust_ the cert, it instead **skips** a security check. Omit `-k` here (and add other options around the CA certificate), or reword to explain what's going on.

```

Triggering this **kubelet** API will request the creation of a checkpoint from
CRI-O. CRI-O requests a checkpoint from your low-level runtime (for example,
`runc`). Seeing that request, `runc` invokes the `criu` tool
to do the actual checkpointing.

Once the checkpointing has finished the checkpoint should be available at
`/var/lib/kubelet/checkpoints/checkpoint-<pod-name>_<namespace-name>-<container-name>-<timestamp>.tar`

You could then use that tar archive to restore the container somewhere else.

> ### Restore a checkpointed container outside of Kubernetes (with CRI-O) {#restore-checkpointed-container-standalone)

Reviewer: Oops
```suggestion
### Restore a checkpointed container outside of Kubernetes (with CRI-O) {#restore-checkpointed-container-standalone}
```

With the checkpoint tar archive it is possible to restore the container outside
of Kubernetes in a sandboxed instance of CRI-O. For better user experience
during restore, I recommend that you use the latest version of CRI-O from the
*main* CRI-O GitHub branch. If you're using CRI-O v1.25, you'll need to
manually create certain directories Kubernetes would create before starting the
container.

The first step to restore a container outside of Kubernetes is to create a pod sandbox
using *crictl*:
```shell
crictl runp pod-config.json
```

Then you can restore the previously checkpointed container into the newly created pod sandbox:
```shell
crictl create <POD_ID> container-config.json pod-config.json
```

Instead of specifying a container image in a registry in `container-config.json`
you'd need specify the path to the checkpoint archive that you created earlier:
```yaml
{
  "metadata": {
      "name": "counter"
  },
  "image":{
      "image": "/var/lib/kubelet/checkpoints/<checkpoint-archive>.tar"
  }
}
```

Next, run `crictl start <CONTAINER_ID>` to start that container, and then a
copy of the previously checkpointed container should be running.

> ### Restore a checkpointed container within of Kubernetes {#restore-checkpointed-container-k8s)

Reviewer: ```suggestion
### Restore a checkpointed container within of Kubernetes {#restore-checkpointed-container-k8s}
```

To restore the previously checkpointed container directly in Kubernetes it is
necessary to convert the checkpoint archive into an image that can be pushed to
a registry.

One possible way to convert the local checkpoint archive consists of the
following steps with the help of *buildah*:
```shell
newcontainer=$(buildah from scratch)
buildah add $newcontainer /var/lib/kubelet/checkpoints/checkpoint-<pod-name>_<namespace-name>-<container-name>-<timestamp>.tar /
buildah config --annotation=io.kubernetes.cri-o.annotations.checkpoint.name=<container-name> $newcontainer
buildah commit $newcontainer checkpoint-image:latest
```

The resulting image is not standardized and only works in combination with CRI-O.
Please consider this image format as pre-alpha. There are ongoing discussions
to standardize the format of checkpoint images like this. Important to remember
is that this not yet standardized image format only works if CRI-O has been started

> with `--enable-criu-support=true --drop-infra-ctr=false`. The security implications

Reviewer: ```suggestion
with `--enable-criu-support=true` and `--drop-infra-ctr=false`. The security implications
```

forensic container checkpointing as implemented in Kubernetes
uses these existing CRIU integrations.

[criu]: https://criu.org/

## Why is it important?

With the help of CRIU and the corresponding integrations it is possible to get
all information and state about a running container on disk for later forensic
analysis. Forensic analysis might be important to inspect a suspicious
container without stopping or influencing it. If the container is really under
attack, the attacker might detect attempts to inspect the container. Taking a
checkpoint and analysing the container in a sandboxed environment offers the
possibility to inspect the container without the original container and maybe
attacker being aware of the inspection.

In addition to the forensic container checkpointing use case, it is also
possible to migrate a container from one node to another node without loosing
the internal state. Especially for stateful containers with long initialization
times restoring from a checkpoint might save time after a reboot or enable much
faster startup times.

## How do I use container checkpointing?

The feature is behind a feature gate, so make sure to enable the

> `ContainerCheckpointRestore` gate before you can use the new feature.

Reviewer 2: Can we link to some official docs about the feature gate here? Is it a kubelet or API one?

Author: It is listed here, but I do not think I can link directly to `ContainerCheckpoint` feature gate description: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/

There is also https://kubernetes.io/docs/reference/node/kubelet-checkpoint-api/ . Do you think linking here would make sense?

Author: Done

The runtime must also support container checkpointing:

* containerd: support is currently under discussion. See containerd
  pull request [#6965][containerd-checkpoint-restore-pr] for more details.

* CRI-O: v1.25 has support for forensic container checkpointing.

[containerd-checkpoint-restore-pr]: https://github.com/containerd/containerd/pull/6965

### Usage example with CRI-O

> To use forensic container checkpointing in combination with CRI-O, CRI-O needs

Reviewer 2: ```suggestion
To use forensic container checkpointing in combination with CRI-O, the runtime needs
```

`/var/lib/kubelet/checkpoints/checkpoint-<pod-name>_<namespace-name>-<container-name>-<timestamp>.tar`

You could then use that tar archive to restore the container somewhere else.

### Restore a checkpointed container outside of Kubernetes (with CRI-O) {#restore-checkpointed-container-standalone}

With the checkpoint tar archive it is possible to restore the container outside
of Kubernetes in a sandboxed instance of CRI-O. For better user experience
during restore, I recommend that you use the latest version of CRI-O from the
*main* CRI-O GitHub branch. If you're using CRI-O v1.25, you'll need to
manually create certain directories Kubernetes would create before starting the
container.

The first step to restore a container outside of Kubernetes is to create a pod sandbox
using *crictl*:
```shell
crictl runp pod-config.json
```

Then you can restore the previously checkpointed container into the newly created pod sandbox:
```shell
crictl create <POD_ID> container-config.json pod-config.json
```

Instead of specifying a container image in a registry in `container-config.json`
you'd need specify the path to the checkpoint archive that you created earlier:

> ```yaml

Reviewer 2: ```suggestion
```json
```

{
  "metadata": {
      "name": "counter"
  },
  "image":{
      "image": "/var/lib/kubelet/checkpoints/<checkpoint-archive>.tar"
  }
}
```

Next, run `crictl start <CONTAINER_ID>` to start that container, and then a
copy of the previously checkpointed container should be running.

### Restore a checkpointed container within of Kubernetes {#restore-checkpointed-container-k8s}

To restore the previously checkpointed container directly in Kubernetes it is
necessary to convert the checkpoint archive into an image that can be pushed to
a registry.

One possible way to convert the local checkpoint archive consists of the

> following steps with the help of *buildah*:

Reviewer 2: Let's link buildah here to its website.

Author: Done
